### DISCUSSION TITLE
Allow pattern matching in catch block exception specifiers

### DISCUSSION THREAD

**Discussion by reflectronic** [dotnet/csharplang#6447] [Language Ideas]
In C#, the exception specifier of a catch block can take one of the following forms:
- `catch (TException)`
- `catch (TException ex)`
- `catch` (no specifier present)

This syntax provides a simple model for defining a catch block's condition. For a long time, this more 'declarative' way of expressing a condition based on an object's 'shape' was unique to exception specifiers. The inflexibility of this model can be worked around through exception filters, which allow an additional arbitrary expression to be evaluated as part of the catch block's condition.

C# has recently invested heavily in new syntax for writing similarly 'declarative' conditions over any kind of object (patterns). Patterns can express a superset of conditions expressible by exception specifiers. It would seem a natural extension of pattern matching and exception filters to _allow pattern matching inside of the exception specifier of a catch block_. 

### Basic proposal
The exception specifier of a catch block is redefined as the _pattern_ of a catch block. The compile-time type of the expression that the pattern is matched against is `Exception`.
```cs
try { result = await client.GetStringAsync(uri); }
catch (HttpRequestException { StatusCode: HttpStatusCode.NotFound }) { ... }
```

The existing syntax forms `catch (TException)` and `catch (TException ex)` can be retconned as uses of the type pattern and the declaration pattern respectively. 

Variable declarations inside of a catch block's pattern can be used inside the catch block.
```cs
try { feed = ReadFeedAsync(uri).Result; }
catch (AggregateException { InnerException: SocketException ex })
{
    Console.WriteLine("Networking error: " + ex.Message);
    Console.WriteLine("Could not retrieve feed.");
}
catch (Exception ex)
{
    Console.WriteLine("Unhandled exception: " + ex);
    throw;
}
```

Like catch blocks today, later catch blocks cannot have been subsumed by earlier ones. 
```cs
try { settings = ReadConfigurationFile(); }
catch (FileNotFoundException { Path: var path }) { ... }
catch (FileNotFoundException and DirectoryNotFoundException) // error, since a value cannot be both
catch (IOException or UnauthorizedAccessException) { ... }
catch (PathTooLongException) { ... } // error, since the previous catch block subsumes this one
catch (10) { ... } // error, since an Exception will never match this pattern
```

Naturally, exception filters can be used together with patterns, creating parity with switch statements/expressions:
```cs
try { data = AccessService(); }
catch (HttpRequestException { StatusCode: >= HttpStatusCode.BadGateway and <= HttpStatusCode.GatewayTimeout }) when (RetryOnTransientError)
{
    await Task.Delay(5000);
    Retry();
}
```

### Problems

A somewhat common pattern with exception filters is as follows:
```cs
catch (Exception ex) when (ex is CommunicationException or TimeoutException) { ... }
```
When the variable `ex` is not needed within the catch block, this proposal simplifies the example nicely:
```cs
catch (CommunicationException or TimeoutException) { ... }
```
However, when `ex` is required, the readability benefit over the exception filter is questionable:
```cs
catch (Exception ex and CommunicationException or TimeoutException) { ... }
```
---
The syntax also does not play particularly well with negated patterns. Intuitively, one might take code like this
```cs
catch (Exception ex) when (ex is not PipelineStoppedException) { ... }
```
and try to simplify it, expecting this code to create a variable `Exception ex` that will never be a `PipelineStoppedException`:
```cs
catch (not PipelineStoppedException ex) { ... }
```
However, patterns do not work this way in C#, and achieving the desired code again requires using the `Exception ex and` workaround.

**Comment by CyrusNajmabadi** [comment]
I really like the concept here. Especially the retconning.  However, I also feel like exceptions are not an important enough area to invest in. The gain over the existing syntax is marginal.

**Reply by HaloFour** [reply]
I personally like the idea of mixing exception handling in with other pattern matching cases.  I look at F# and it feels jarring that to match on the result of a method that might throw you need to nest two different but incredibly similar constructs.  Some way to combine `try` with `switch` might be interesting, then apply something like a `catch` pattern to handle the exceptions that arise from the expression:

I don't like this syntax, but this being the general idea:

```cs
string result = try SomeExpression() switch {
    SomePattern { SomeProperty: string value } => value,
    default => "Foo",
    catch HttpRequestException { StatusCode: HttpStatusCode.NotFound } => "Bar",
    catch HttpRequestException ex => ex.StatusCode.ToString(),
    catch (CommunicationException or TimeoutException) => "Baz",
    catch => throw
};
```

I think something like that would also satisfy the numerous proposals to "simplify" the syntax of `try` statements that have been suggested on this repo, to some degree.  Although I think there'd still be concerns as to how much easier it might make it to "ignore" exceptions, or what exhaustiveness checking now means especially when you mix values and thrown exceptions.
